walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,60 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    //Approach: start form start of day, and add available timeranges as we traverse through the events' timeranges
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<TimeRange> validTimeRanges = new ArrayList();
+        long duration = request.getDuration();
+        int start = 0;
+        int end = 0;
+        int lastEnd = 0;
+
+        // Get all the event time ranges and sort them by the start time. 
+        // Only add to the list if at least one attendee from the event is 
+        //a required attendee in the meeting request. (Otherwise, the event is irrelevant.)
+        List<TimeRange> eventTimeRanges = new ArrayList();
+        for (Event event: events) {
+            for (String attendee: event.getAttendees()){
+                if (request.getAttendees().contains(attendee)){
+                    eventTimeRanges.add(event.getWhen());
+                    break;
+                }
+            }
+        }
+        Collections.sort(eventTimeRanges, TimeRange.ORDER_BY_START);
+        
+        // Check for invalid input of duration.
+        if(duration > TimeRange.WHOLE_DAY.duration()) {
+            return validTimeRanges;
+        }
+        // If there are no scheduled events, the meeting can be scheduled at any time. 
+        if (eventTimeRanges.size() == 0){
+            validTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,TimeRange.END_OF_DAY,true));
+        }else {
+             // Loop through all events.
+            for (int i = 0; i < eventTimeRanges.size(); i++) {
+                TimeRange eventTimeRange = eventTimeRanges.get(i);
+                end = eventTimeRange.start();
+                if (end-start >= duration) {
+                    validTimeRanges.add(TimeRange.fromStartEnd(start,end,false));
+                }
+                start = eventTimeRange.end();
```

Comment:
I think you only need to update start if you also update lastEnd

Reply:
Yes! So in the next PR I decided to just update start instead, lastEnd is kind of redundant.

Reply:
sounds good.